jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/thread/QueuedThreadPool.java

```diff
@@ -789,47 +802,68 @@ public void execute(Runnable job)
     {
         // Determine if we need to start a thread, use and idle thread or just queue this job
         int startThread;
+        long counts = _counts.get();
         while (true)
         {
-            // Get the atomic counts
-            long counts = _counts.get();
-
             // Get the number of threads started (might not yet be running)
             int threads = AtomicBiInteger.getHi(counts);
             if (threads == Integer.MIN_VALUE)
+            {
+                reject(job);
                 throw new RejectedExecutionException(job.toString());
+            }
 
             // Get the number of truly idle threads. This count is reduced by the
             // job queue size so that any threads that are idle but are about to take
             // a job from the queue are not counted.
             int idle = AtomicBiInteger.getLo(counts);
 
-            // Start a thread if we have insufficient idle threads to meet demand
+            // Start a thread if we have not enough idle threads to meet demand
             // and we are not at max threads.
             startThread = (idle <= 0 && threads < _maxThreads) ? 1 : 0;
 
             // Add 1|0 or 0|-1 to counts depending upon the decision to start a thread or not;
             // idle can become negative which means there are queued tasks.
-            if (!_counts.compareAndSet(counts, threads + startThread, idle + startThread - 1))
-                continue;
-
-            break;
+            long nextCounts = AtomicBiInteger.encode(threads + startThread, idle + startThread - 1);
+            if (_counts.compareAndSet(counts, nextCounts))
+            {
+                counts = nextCounts;
+                break;
+            }
+            counts = _counts.get();
         }
 
-        if (!_jobs.offer(job))
+        // Start threads if they are needed
+        boolean started = startThread > 0;
+        while (startThread-- > 0)
+            startThread();
+
+        // Return if we were able to queue the job
+        if (_jobs.offer(job))
+            return;
+
+        // The queue is full, so let's spin until we see some started threads
+        if (started)
         {
-            // reverse our changes to _counts.
-            if (addCounts(-startThread, 1 - startThread))
-                LOG.warn("{} rejected {}", this, job);
-            throw new RejectedExecutionException(job.toString());
+            while (true)
+            {
+                // Spin wait to allow threads to start
+                Thread.onSpinWait();
+                if (_counts.get() != counts)
```

Comment:
I think this is a good basic approach, that if the queue is full we spin waiting for any started threads to have been started.

HOWEVER! This is the wrong test for that. Counts will only be updated if a Runner completes running a job or if more jobs are queued (and they may be spinning as well).

We don't really have anything we can spin on to ensure a thread is started.

The only other approach I can think of is to start a special Runner that has a reference to the job.  When that Runner takes a job from the queue it will immediately offer the job... but that may still race with other threads adding to the queue.  Perhaps it just keeps the job reference and tries to queue it after every job it takes until it can and then nulls it.

Let me try that.